File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    var answer = 0;

    function calTime(diff, level, time_cur, time_prev) {
        if (diff <= level) {
            return time_cur;
        }
        return (diff - level) * (time_prev + time_cur) + time_cur;
    }

    let left = 1, right = limit, mid, tmp;
    
    while (left <= right) {
        mid = Math.floor((left + right) / 2);
        tmp = times[0];
        
        for (let i = 1; i < diffs.length; i++) {
            tmp += calTime(diffs[i], mid, times[i], times[i - 1]);
        }
        
        if (tmp <= limit) {
            answer = mid;
            right = mid - 1;
        } else {
            left = mid + 1; 
        }
    }

    return answer;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):
    answer = 0
    left, right = 1, limit
    
    while (left <= right):
        mid = (left + right)//2
        tmp = times[0]
        
        for i in range(1, len(diffs)):
            if mid < diffs[i] :
                tmp += (times[i] + times[i-1]) * (diffs[i] - mid) + times[i]
            else:
                tmp += times[i]
                
        if tmp <= limit:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1
    return answer